src/main/java/com/jozufozu/flywheel/backend/engine/uniform/FrameUniforms.java

```diff
@@ -56,39 +88,109 @@ public void write(long ptr) {
+			cleanViewProjectionPrev.set(cleanViewProjection);
+		}
 		ptr = writeMatrices(ptr);
-
-		ptr = writeCamera(ptr, camX, camY, camZ, camera);
+		viewPrev.set(view);
+		projectionPrev.set(projection);
+		viewProjectionPrev.set(viewProjection);
+		cleanProjectionPrev.set(cleanProjection);
+		cleanViewProjectionPrev.set(cleanViewProjection);
+
+		// last values for camera
+		if (!lastInit) {
+			cameraPositionPrev.set(camX, camY, camZ);
+			cameraLookPrev.set(camera.getLookVector());
+			cameraRotPrev.set(camera.getXRot(), camera.getYRot());
+		}
+		ptr = writeCamera(ptr, camX, camY, camZ);
+		cameraPositionPrev.set(camX, camY, camZ);
+		cameraLookPrev.set(camera.getLookVector());
+		cameraRotPrev.set(camera.getXRot(), camera.getYRot());
 
 		var window = Minecraft.getInstance()
 				.getWindow();
-		ptr = writeVec2(ptr, window.getWidth(), window.getHeight());
+		ptr = Uniforms.writeVec2(ptr, window.getWidth(), window.getHeight());
 
 		// default line width: net.minecraft.client.renderer.RenderStateShard.LineStateShard
 		MemoryUtil.memPutFloat(ptr, Math.max(2.5F, (float) window.getWidth() / 1920.0F * 2.5F));
 		ptr += 4;
 
+		MemoryUtil.memPutFloat(ptr, (float) window.getWidth() / (float) window.getHeight());
+		ptr += 4;
+
+		MemoryUtil.memPutFloat(ptr, Minecraft.getInstance().gameRenderer.getDepthFar());
+		ptr += 4;
+
 		MemoryUtil.memPutInt(ptr, getConstantAmbientLightFlag(context));
 		ptr += 4;
 
 		ptr = writeTime(ptr);
 
+		ptr = writeCameraIn(ptr);
+
 		MemoryUtil.memPutInt(ptr, debugMode);
+
+		lastInit = true;
 	}
 
 	private long writeMatrices(long ptr) {
-		MatrixMath.writeUnsafe(viewProjection, ptr);
-		MatrixMath.writeUnsafe(viewProjection.invert(viewProjectionInverse), ptr + 64);
-		return ptr + 128;
+		MatrixMath.writeUnsafe(view, ptr);
+		MatrixMath.writeUnsafe(view.invert(viewInverse), ptr + 64);
+		MatrixMath.writeUnsafe(viewPrev, ptr + 64 * 2);
+		MatrixMath.writeUnsafe(projection, ptr + 64 * 3);
+		MatrixMath.writeUnsafe(projection.invert(projectionInverse), ptr + 64 * 4);
+		MatrixMath.writeUnsafe(projectionPrev, ptr + 64 * 5);
+		MatrixMath.writeUnsafe(viewProjection, ptr + 64 * 6);
+		MatrixMath.writeUnsafe(viewProjection.invert(viewProjectionInverse), ptr + 64 * 7);
+		MatrixMath.writeUnsafe(viewProjectionPrev, ptr + 64 * 8);
+		MatrixMath.writeUnsafe(cleanProjection, ptr + 64 * 9);
+		MatrixMath.writeUnsafe(cleanProjection.invert(cleanProjectionInverse), ptr + 64 * 10);
+		MatrixMath.writeUnsafe(cleanProjectionPrev, ptr + 64 * 11);
+		MatrixMath.writeUnsafe(cleanViewProjection, ptr + 64 * 12);
+		MatrixMath.writeUnsafe(cleanViewProjection.invert(cleanViewProjectionInverse), ptr + 64 * 13);
+		MatrixMath.writeUnsafe(cleanViewProjectionPrev, ptr + 64 * 14);
+		return ptr + 64 * 15;
 	}
 
-	private static long writeCamera(long ptr, float camX, float camY, float camZ, Camera camera) {
-		ptr = writeVec3(ptr, camX, camY, camZ);
+	private void setupCleanMatrices(PoseStack stack, Camera camera, float partialTicks) {
```

Comment:
I feel like manually setting up the matrices could cause an issue where another mod directly modifies the original matrix using a `GameRenderer` mixin and these clean matrices do not include those additional transformations.

Reply:
I guess this depends on what exactly the definition of 'clean' is. For this, I had just been using Canvas as my reference for how to implement this. With the current definition, a clean projection matrix is purely the perspective matrix. Furthermore, injecting into `GameRenderer` requires using a local capture.